src/cm/communication_manager.rs

```diff
@@ -98,13 +360,277 @@ impl CommunicationManager {
+                    None => null_mut(),
+                },
+                match dst_addr {
+                    Some(addr) => OsSocketAddr::from(addr).as_mut_ptr(),
+                    None => null_mut(),
+                },
+                timeout_ms,
+            )
+        };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve address {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn resolve_route(&self, timeout_ms: i32) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+        let ret = unsafe { rdma_resolve_route(cm_id.as_mut(), timeout_ms) };
+
+        if ret < 0 {
+            return Err(format!("Failed to resolve route {ret}"));
+        }
+
+        Ok(())
+    }
+
+    pub fn listen(&self, backlog: i32) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+        let ret = unsafe { rdma_listen(cm_id.as_mut(), backlog) };
 
         if ret < 0 {
             return Err(format!("Failed to listen {ret}"));
         }
 
         Ok(())
     }
+
+    pub fn get_device_context(&self) -> Result<ManuallyDrop<DeviceContext>, String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+
+        unsafe {
+            Ok(ManuallyDrop::new(DeviceContext {
+                context: (*cm_id.as_ptr()).verbs,
+            }))
+        }
+    }
+
+    pub fn connect(&self, conn_param: &mut ConnectionParameter) -> Result<(), String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+        let ret = unsafe { rdma_connect(cm_id.as_ptr(), &mut conn_param.0) };
+
+        if ret < 0 {
+            return Err(format!("Failed to connect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn disconnect(&self) -> Result<(), String> {
+        let cm_id = unsafe { *self.cm_id.get() };
+        let ret = unsafe { rdma_disconnect(cm_id.as_ptr()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to disconnect {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn accept(&self, conn_param: Option<&mut ConnectionParameter>) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+
+        let ret = match conn_param {
+            Some(param) => unsafe { rdma_accept(cm_id.as_mut(), &mut param.0) },
+            None => unsafe { rdma_accept(cm_id.as_mut(), null_mut()) },
+        };
+
+        if ret < 0 {
+            return Err(format!("Failed to accept {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn establish(&self) -> Result<(), String> {
+        let cm_id = unsafe { &mut *self.cm_id.get() };
+        let ret = unsafe { rdma_establish(cm_id.as_mut()) };
+
+        if ret < 0 {
+            return Err(format!("Failed to establish {:?}", io::Error::last_os_error()));
+        }
+
+        Ok(())
+    }
+
+    pub fn create_qp(&self) -> Result<(), String> {
```

Comment:
I suggest not to implement bindings for the `rdma_create_qp`-related APIs since they are quite subtle to use. For example, these APIs are event not thread-safe (https://github.com/linux-rdma/rdma-core/pull/1182). Users are encouraged to use the self-create QP way to manage QPs, which can take advantage of our QP factory for creating QPs. This argument applies for the following methods: `create_qp`, `destroy_qp`, `bind_qp`, `qp`.